domain/consensus/processes/blockprocessor/validateandinsertblock.go

```diff
@@ -4,48 +4,111 @@ import (
 	"github.com/kaspanet/kaspad/domain/consensus/model"
 	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
 	"github.com/kaspanet/kaspad/domain/consensus/ruleerrors"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/hashserialization"
 	"github.com/pkg/errors"
 )
 
-func (bp *blockProcessor) validateAndInsertBlock(block *externalapi.DomainBlock) error {
-	err := bp.validateBlock(block)
+func (bp *blockProcessor) validateAndInsertBlock(block *externalapi.DomainBlock, headerOnly bool) error {
+	hash := hashserialization.HeaderHash(block.Header)
+	if headerOnly && len(block.Transactions) != 0 {
+		return errors.Errorf("block %s contains transactions while validating in header only mode", hash)
+	}
+
+	err := bp.checkBlockStatus(hash, headerOnly)
+	if err != nil {
+		return err
+	}
+
+	err = bp.validateBlock(block, headerOnly)
 	if err != nil {
 		bp.discardAllChanges()
 		return err
 	}
 
+	hasHeader, err := bp.hasHeader(hash)
+	if err != nil {
+		return err
+	}
+
+	if !hasHeader {
+		err = bp.reachabilityTree.AddBlock(hash)
+		if err != nil {
+			return err
+		}
+
+		err = bp.headerTipsManager.AddBlock(hash)
+		if err != nil {
+			return err
+		}
+	}
+
+	if headerOnly {
+		bp.blockStatusStore.Stage(hash, model.StatusHeaderOnly)
+	} else {
+		bp.blockStatusStore.Stage(hash, model.StatusUTXOPendingVerification)
+	}
+
 	// Block validations passed, save whatever DAG data was
 	// collected so far
 	err = bp.commitAllChanges()
 	if err != nil {
 		return err
 	}
 
-	// Attempt to add the block to the virtual
-	err = bp.consensusStateManager.AddBlockToVirtual(block.Hash)
+	if !headerOnly {
```

Comment:
Instead, do if headerOnly { return nil } (this also avoids running commitAllChanges again)